src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs

```diff
@@ -0,0 +1,266 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.SolutionSize;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Storage
+{
+    /// <summary>
+    /// A service that enables storing and retrieving of information associated with solutions,
+    /// projects or documents across runtime sessions.
+    /// </summary>
+    internal abstract partial class AbstractPersistentStorageService : IPersistentStorageService2
+    {
+        private readonly IOptionService _optionService;
+        private readonly IPersistentStorageLocationService _locationService;
+        private readonly SolutionSizeTracker _solutionSizeTracker;
+
+        /// <summary>
+        /// This lock guards all mutable fields in this type.
+        /// </summary>
+        private readonly object _lock = new object();
+        private ReferenceCountedDisposable<IPersistentStorage> _currentPersistentStorage;
+        private SolutionId _currentPersistentStorageSolutionId;
+        private bool _subscribedToLocationServiceChangeEvents;
+
+        protected AbstractPersistentStorageService(
+            IOptionService optionService,
+            IPersistentStorageLocationService locationService,
+            SolutionSizeTracker solutionSizeTracker)
+        {
+            _optionService = optionService;
+            _locationService = locationService;
+            _solutionSizeTracker = solutionSizeTracker;
+        }
+
+        protected abstract string GetDatabaseFilePath(string workingFolderPath);
+        protected abstract bool TryOpenDatabase(Solution solution, string workingFolderPath, string databaseFilePath, out IPersistentStorage storage);
+        protected abstract bool ShouldDeleteDatabase(Exception exception);
+
+        public IPersistentStorage GetStorage(Solution solution)
+            => GetStorage(solution, checkBranchId: true);
+
+        public IPersistentStorage GetStorage(Solution solution, bool checkBranchId)
+        {
+            if (!DatabaseSupported(solution, checkBranchId))
+            {
+                return NoOpPersistentStorage.Instance;
+            }
+
+            lock (_lock)
+            {
+                // Do we already have storage for this?
+                if (solution.Id == _currentPersistentStorageSolutionId)
+                {
+                    // We do, great
+                    return PersistentStorageReferenceCountedDisposableWrapper.AddRefrenceCountToAndCreateWrapper(_currentPersistentStorage);
+                }
+
+                if (!SolutionSizeAboveThreshold(solution))
+                {
+                    return NoOpPersistentStorage.Instance;
+                }
+
+                string workingFolder = _locationService.TryGetStorageLocation(solution.Id);
```

Comment:
so it no longer support case where solution is closed and new solution is opened, pending async task that hold onto last solution getting persistent service support. only new solution work.

one usage case before was that, when solution is closed but VS alive, pending task can keep saving its state to db so that it can be picked up next time. might not a bit deal with most people just kill VS. rather than close and reopen different solution.

Reply:
@heejaechang Does this still work after your change in #24458? UnregisterPrimarySolution would get called once we closed down, and we'd dispose the storage and be done with it. There's no lifetime tracking we have for "somebody might still have a solution and they might ask for the storage sometime far later." And after somebody closed the solution they might have already moved it on disk, switched git branches, etc?

Reply:
I think we talked about this offline?